src/lib/scheduler/node_queue_scheduler.cpp

```diff
@@ -63,7 +75,23 @@ void NodeQueueScheduler::begin() {
 }
 
 void NodeQueueScheduler::wait_for_all_tasks() {
+  auto progressless_loop_count = size_t{0};
+  auto previous_finished_task_count = TaskID::base_type{0};
+
+  // To check if the system is still processing incoming jobs, we store the previous task count and loop-wait until no
+  // new jobs are created anymore.
+  auto previous_task_count = TaskID::base_type{_task_counter.load()};
+
   while (true) {
+    if (_task_counter > previous_task_count) {
+      // System is still processing new tasks (can happen when, e.g., currently running tasks schedule new tasks):
+      // loop-wait until task counter is stable (this check can still fail in edge cases, but we make the simple
+      // assumption that nobody calls wait_for_all_tasks() if there is still significant processing ongoing).
+      std::this_thread::sleep_for(std::chrono::milliseconds(10));
+      previous_task_count = _task_counter;
+      continue;
+    }
```

Comment:
Talking about the instability: Does it make sense to increase the sleep time or ensure the count is stable for, e.g., three loops? Also, is there a reason to sleep before the assignment? This means we check, sleep, assign, directly check again, ... Does it make sense to check, assign, sleep, check again, ...?

Reply:
I thought about that as well. The problem is that it very regularly happens that the first few loops look "stable" and then the change happens. We could always first wait for 10 ms and then check for progress (also just a guess in the end), but that would slow down many test cases.

Assigning after the sleep was wrong. Thanks.

Reply:
Okay, guess then I'm fine